terraform/README.md

```diff
@@ -1 +1,98 @@
-# TODO
\ No newline at end of file
+# XMTPD Terraform Modules
+
+**⚠️ Experimental:** This software is in early development. Expect frequent changes and unresolved issues.
+
+`xmtpd` (XMTP daemon) is an experimental version of XMTP node software. It is **not** the node software that currently forms the XMTP network.
+
+This repository includes Terraform modules that can be used as part of a new or existing Terraform plan to add `xmtpd` nodes to your infrastructure.
+
+Currently only AWS is supported with plans for future support for Google Cloud and other hosting providers.
+
+## Pre-requisites
+
+1. An AWS account
+2. An IAM user in that account with privileges to add and remove a wide range of infrastructure (the built-in Administrator policy will work, or you can make a more tailored policy if desired)
+3. AWS credentials for that user (`AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`)
+4. An Alchemy account with an active subscription, and an app in that account with access to the following blockchains:
+   - Ethereum
+   - Base
+   - Arbitrum One
+   - OP Mainnet
+   - Polygon Mainnet
+   - ZKSync Mainnet
+   - Linea Mainnet
+   - XMTP Sepolia
+5. A private key for your node, which can be generated using the instructions in the `xmtpd` repo [here](https://github.com/xmtp/xmtpd/blob/main/doc/onboarding.md)
+
+## Register a domain
+
+1. Choose a new domain name for your node
+2. Register the domain in [Route53](https://us-east-1.console.aws.amazon.com/route53/v2/home?region=us-east-2#Dashboard)
+3. Wait for the registration to complete (may take 15-20 minutes)
+
+## Deploying into a new VPC
+
+### Set up terraform variables
+
+#### With Terraform Cloud
+
+1. Create a workspace named `xmtp-testnet`
+2. Add the following variables to the workspace
+
+```terraform
+verifier_chain_rpc_urls = {
+    chain_rpc_1 = $ETHEREUM_RPC_URL
+    chain_rpc_8453 = $BASE_RPC_URL
+    chain_rpc_42161 = $ARBITRUM_ONE_RPC_URL
+    chain_rpc_10 = $OP_MAINNET_RPC_URL
+    chain_rpc_324 = $POLYGON_MAINNET_RPC_URL
+    chain_rpc_59144 = $LINEA_MAINNET_RPC_URL
+}
+
+nodes_contract_address = "0x390D339A6C0Aa432876B5C898b16287Cacde2A0A"
+messages_contract_address = "0x162f2d4d96565437F47bfB7a0BF8AC4FF481Bbf6"
+identity_updates_contract_address = "0x00e92e15AB0D7d3aA5c76bceCcE675DcAf311189"
```

Comment:
They're the ones in use in our current testnet. You can copy over for now, although we should come up with a central place to hold these values so we only have to update one place instead of 6 when we eventually rewrite these contracts.

Reply:
We could adopt a proxy + create2 approach for contract deployments.
Making the contracts upgradable through a proxy would guarantee that, once deployed, the address won't ever change.
CREATE2 is a optional; it's an EVM opcode that would give us the capability of predicting the address where a contract is going to be deployed by using a salt. This could be used in the future if we consider migrating between blockchains and would want to preserve the address for some reason.

Reply:
I like that idea. We would still need to update some config if the shape of the events ever changed, since the ABI parsing lives in the code and would need to be updated at the same time.

But most contract changes don't change the event structure.